### PR TITLE
Add Vertex AI Featurestores

### PR DESCRIPTION
**Pull request by upodroid**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/9264
Part of: https://github.com/hashicorp/terraform-provider-google/issues/9298

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_vertex_ai_featurestore`
```
```release-note:new-resource
`google_vertex_ai_featurestore_entitytype`
```

### CODE REVIEW

**Comment by upodroid** [comment]
Is there an easy way of inferring project and region from a self_link style field such as featurestore?

Instead of 

```
resource "google_vertex_ai_featurestore_entitytype" "featurestore" {
  provider = google-beta
  region = "us-central1"
  project = "foo"
  name     = "terraform%{random_suffix}"
  labels = {
    foo = "bar"
  }
  featurestore = "foobar"
  monitoring_config {
    snapshot_analysis {
      disabled = false
      monitoring_interval = "86400s"
    }
  }
}
```

it becomes

```
resource "google_vertex_ai_featurestore_entitytype" "featurestore" {
  provider = google-beta
  name     = "terraform%{random_suffix}"
  labels = {
    foo = "bar"
  }
  featurestore = "projects/{{project}}/locations/{{region}}/featurestores/{{featurestore}}/entityTypes/{{name}}"
  monitoring_config {
    snapshot_analysis {
      disabled = false
      monitoring_interval = "86400s"
    }
  }
}
```

If this was handwritten, it would be easy to do but this requires resource.erb modifications and my last major change to it didn't go so well.

**Comment by c2thorn** [comment]
> Is there an easy way of inferring project and region from a self_link style field such as featurestore?

You should be able to do so by adding an `encoder` override which does the logic. See `resourcePubsubLiteSubscriptionEncoder` derived from https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/pubsublite/terraform.yaml#L30

**Review by c2thorn — APPROVED** [review]
Looks good as it is now. I'll wait for your response on whether or not you want to add the encoder before merging. Otherwise, I'll merge tomorrow.

File: mmv1/products/vertexai/api.yaml
Comment (upodroid) [inline mmv1/products/vertexai/api.yaml:183]: If I convert this to `'{{featurestore}}/entityTypes/{{name}}'`, it won't work thanks to has_project variable in resource.erb

What I want to do is:
- Remove project and region from schema. 
- A custom pre_create, pre_update and pre_delete snippet with the following code:

```
if v, ok := d.GetOk("featurestore"); ok {
	re := regexp.MustCompile("projects/([a-zA-Z0-9-]*)/(?:locations|regions)/([a-zA-Z0-9-]*)")
	switch {
	case re.MatchString(selfLink):
		if res := re.FindStringSubmatch(selfLink); len(res) == 3 && res[1] != "" {
			project = res[1]
			region = res[2]
		}
	}
}
```

Comment (c2thorn) [reply mmv1/products/vertexai/api.yaml:183]: Hi @upodroid,
Looking more into this, I think the data catalog entryGroup + entry relationship is similar to this scenario. Take a look: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/datacatalog/api.yaml#L80

We could similarly separate out the base_url and create_url and add a `entityTypeId` url param. And then make `name` an output field and set the self_link to that. Looks like doing so makes import functionality tricky, so there is a custom import override as well: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/mmv1/products/datacatalog/terraform.yaml#L79

Comment (upodroid) [reply mmv1/products/vertexai/api.yaml:183]: That won't work. DatacatalogEntry returns the Entry instead of an Operation. I tried that earlier

```
REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  16⬇  9✔  3✎  1⚑  $    make test
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
go generate  ./...
go test  -timeout=30s $(go list ./...)
# github.com/hashicorp/terraform-provider-google-beta/google-beta
google-beta/resource_vertex_ai_featurestore_entitytype.go:172:45: not enough arguments in call to vertexAIOperationWaitTimeWithResponse
        have (*Config, map[string]interface {}, *map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, *map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:298:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:336:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
note: module requires Go 1.16
# github.com/hashicorp/terraform-provider-google-beta/google-beta [github.com/hashicorp/terraform-provider-google-beta/google-beta.test]
google-beta/resource_vertex_ai_featurestore_entitytype.go:172:45: not enough arguments in call to vertexAIOperationWaitTimeWithResponse
        have (*Config, map[string]interface {}, *map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, *map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:298:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
google-beta/resource_vertex_ai_featurestore_entitytype.go:336:33: not enough arguments in call to vertexAIOperationWaitTime
        have (*Config, map[string]interface {}, string, string, time.Duration)
        want (*Config, map[string]interface {}, string, string, string, time.Duration)
note: module requires Go 1.16
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta [build failed]
make: *** [test] Error 2
```

One of the reasons I don't like the encoder is the potential issues with multi regional terraform configs. 

```
provider "google-beta" {
 region = "europe-north1"
}
resource "google_vertex_ai_featurestore" "featurestore" {
  provider = google-beta
  name     = "terraform%{random_suffix}"
  region   = "us-central1"
}

resource "google_vertex_ai_featurestore_entitytype" "featurestore" {
  provider = google-beta
  name     = "terraform%{random_suffix}"
  featurestore = google_vertex_ai_featurestore.featurestore.name
}
```

Comment (upodroid) [reply mmv1/products/vertexai/api.yaml]: nvm, my last commit fixed this issue.

**Comment by upodroid** [comment]
This is ready to be merged now.

Came across this while cleaning up Featurestores, which is fixed in the last commit:

```
 REDACTED  MCW0CDP3YY  ~  $  http DELETE https://us-central1-aiplatform.googleapis.com/v1beta1/projects/REDACTED/locations/us-central1/featurestores/terraformxl34nkqltq "Authorization: Bearer $(gcloud auth print-access-token)"
HTTP/1.1 400 Bad Request
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Cache-Control: private
Content-Encoding: gzip
Content-Type: application/json; charset=UTF-8
Date: Thu, 01 Jul 2021 21:32:03 GMT
Server: ESF
Transfer-Encoding: chunked
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 0

{
    "error": {
        "code": 400,
        "message": "There are entity types in the featurestore. Please delete the child entity types or set `force` in the request to explicitly cascade delete.",
        "status": "FAILED_PRECONDITION"
    }
}
```

**Comment by c2thorn** [comment]
> Came across this while cleaning up Featurestores, which is fixed in the last commit:
I see, so you added `force_destroy` as a virtual field. Makes sense.

File: mmv1/api/async.rb
Comment (c2thorn) [inline mmv1/api/async.rb:69]:
```suggestion
    # If true, include project as an argument to OperationWaitTime.
```
Our linter doesn't like trailing whitespace in ruby

**Comment by upodroid** [comment]
Not sure what is wrong with generate-diffs but i'm able to generate the beta provider locally.

**Comment by c2thorn** [comment]
@upodroid I believe it just needs to be rebased

**Pull request merged**
